main.go

```diff
@@ -1,152 +1,105 @@
+		logger.SetLevel(hclog.Debug)
+	}
 
-	if *version {
-		v, err := getVersion()
+	if *selfVersion {
+		v, err := version.GetVersion()
 		if err != nil {
-			log.Fatalf("failed to print version, err: %+v", err)
+			return fmt.Errorf("failed to print version, err: %w", err)
 		}
 		// print the version and exit
-		fmt.Printf("%s\n", v)
-		os.Exit(0)
+		logger.Info(v)
+		return nil
 	}
 
+	logger.Info("Creating new gRPC server")
 	server := grpc.NewServer()
 
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, syscall.SIGTERM, syscall.SIGINT)
 	go func() {
 		sig := <-c
-		log.Infof("Caught signal %s, shutting down", sig)
+		logger.Info(fmt.Sprintf("Caught signal %s, shutting down", sig))
 		server.GracefulStop()
 	}()
 
-	listener, err := net.Listen("unix", *endpoint)
+	listener, err := listen(logger)
 	if err != nil {
-		log.Fatalf("Failed to listen on unix socket at %s: %v", *endpoint, err)
+		return err
 	}
 	defer listener.Close()
-	log.Infof("Listening on %s", *endpoint)
+	logger.Info(fmt.Sprintf("Listening on %s", *endpoint))
 
-	s := &ProviderServer{}
+	s := &providerserver.ProviderServer{
+		Logger: logger,
+	}
 	pb.RegisterCSIDriverProviderServer(server, s)
 
 	err = server.Serve(listener)
 	if err != nil {
-		log.Fatal(err)
+		return fmt.Errorf("error running gRPC server: %w", err)
 	}
-}
-
-func HandleRequest(ctx context.Context, attributes, secrets, permission, targetPath string) (map[string]int, error) {
-	var attrib, secret map[string]string
-	var filePermission os.FileMode
 
-	err := json.Unmarshal([]byte(attributes), &attrib)
-	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal attributes, err: %v", err)
-	}
-	err = json.Unmarshal([]byte(secrets), &secret)
-	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal secrets, err: %v", err)
-	}
-	err = json.Unmarshal([]byte(permission), &filePermission)
-	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal file permission, err: %v", err)
-	}
-
-	provider, err := NewProvider()
-	if err != nil {
-		return nil, fmt.Errorf("[error] : %v", err)
-	}
-
-	versions, err := provider.MountSecretsStoreObjectContent(ctx, attrib, secret, targetPath, filePermission)
-	if err != nil {
-		return nil, fmt.Errorf("[error] : %v", err)
-	}
-
-	log.Info("Successfully handled mount request")
-
-	return versions, nil
+	return nil
 }
 
-// setupLogger sets up hooks to redirect stdout and stderr
-func setupLogger() {
-	log.SetOutput(ioutil.Discard)
-
-	// set log level
-	log.SetLevel(log.InfoLevel)
-	if *debug {
-		log.SetLevel(log.DebugLevel)
+func listen(logger hclog.Logger) (net.Listener, error) {
+	// Because the unix socket is created in a host volume (i.e. persistent
+	// storage), it can persist from previous runs if the pod was not terminated
+	// cleanly. Check if we need to clean up before creating a listener.
+	_, err := os.Stat(*endpoint)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("failed to check for existence of unix socket: %w", err)
```

Comment:
Should we also warn using the logger here?

Reply:
Same as above, `main()` will log this as an error-level log before the provider exits with exit code 1.